core/templates/components/graph-services/graph-layout.service.spec.ts

```diff
@@ -0,0 +1,1251 @@
+      },
+      State4: {
+        depth: 1,
+        offset: 3,
+        reachable: true,
+        x0: 470.925,
+        y0: 81.19999999999999,
+        xLabel: 527.625,
+        yLabel: 98.00000000000001,
+        id: 'State4',
+        label: 'State4',
+        height: 33.6,
+        width: 113.40000000000002,
+        reachableFromEnd: false
+      }
+    };
+    let nodeDataWithPositionValueInPixel3 = {
+      State1: {
+        depth: 1,
+        offset: 0,
+        reachable: true,
+        x0: 45.675000000000004,
+        y0: 88.2,
+        xLabel: 102.375,
+        yLabel: 105,
+        id: 'State1',
+        label: 'State1',
+        height: 33.6,
+        width: 113.40000000000002,
+        reachableFromEnd: false
+      },
+      State2: {
+        depth: 1,
+        offset: 1,
+        reachable: true,
+        x0: 187.42500000000004,
+        y0: 88.2,
+        xLabel: 244.125,
+        yLabel: 105,
+        id: 'State2',
+        label: 'State2',
+        height: 33.6,
+        width: 113.40000000000002,
+        reachableFromEnd: false
+      },
+      State3: {
+        depth: 1,
+        offset: 2,
+        reachable: true,
+        x0: 329.17500000000007,
+        y0: 88.2,
+        xLabel: 385.875,
+        yLabel: 105,
+        id: 'State3',
+        label: 'State3',
+        height: 33.6,
+        width: 113.40000000000002,
+        reachableFromEnd: false
+      },
+      Introduction: {
+        depth: 0,
+        offset: 0,
+        reachable: true,
+        x0: 45.675000000000004,
+        y0: 32.2,
+        xLabel: 102.375,
+        yLabel: 49,
+        id: 'Introduction',
+        label: 'Introduction',
+        height: 33.6,
+        width: 113.40000000000002,
+        reachableFromEnd: false
+      },
+      State4: {
+        depth: 1,
+        offset: 3,
+        reachable: true,
+        x0: 470.925,
+        y0: 88.2,
+        xLabel: 527.625,
+        yLabel: 105,
+        id: 'State4',
+        label: 'State4',
+        height: 33.6,
+        width: 113.40000000000002,
+        reachableFromEnd: false
+      },
+      State5: {
+        depth: 2,
+        offset: 1,
+        reachable: true,
+        x0: 187.42500000000004,
+        y0: 144.2,
+        xLabel: 244.125,
+        yLabel: 161,
+        id: 'State5',
+        label: 'State5',
+        height: 33.6,
+        width: 113.40000000000002,
+        reachableFromEnd: false
+      }
+    };
+    // The expectedGraphBoundaries are calculated from the x0 and y0 values from
+    // nodeDataWithPositionValueInPixel, where leftEdge is
+    // minimum(nodeDataWithPositionValueInPixel[nodeId].x0 - BORDER_PADDING)
+    // of all nodes, and rightEdge is
+    // maximum(nodeDataWithPositionValueInPixel[nodeId].x0 + BORDER_PADDING +
+    // nodeDataWithPositionValueInPixel[nodeId].width) of all nodes. Similarly,
+    // bottomEdge and topEdge are calculated using y0 and node height.
+    let expectedGraphBoundaries1 = sgls.getGraphBoundaries(
```

Comment:
This seems like actual rather than expected? You're using a method from the service being tested.

Normally expected stuff is hardcoded.

Reply:
Done